tests/replacer/test_single_nodes.py

```diff
@@ -0,0 +1,318 @@
+            ],
+            "storage-0-0": [LOCAL_QUERY],
+            "storage-0-1": [],
+            "storage-1-0": [LOCAL_QUERY],
+            "storage-1-1": [],
+            "storage-2-0": [LOCAL_QUERY],
+            "storage-2-1": [],
+        },
+        id="Replacements through storage nodes",
+    ),
+    pytest.param(
+        "",
+        {
+            "query_node": [
+                "SELECT count() FROM errors_dist FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'",
+                DIST_QUERY,
+            ]
+        },
+        id="Replacements through query node",
+    ),
+    pytest.param(
+        "[10]",
+        {
+            "query_node": [
+                "SELECT count() FROM errors_dist FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'",
+                DIST_QUERY,
+            ]
+        },
+        id="Replacements through query node. Wrong project",
+    ),
+]
+
+COUNT_QUERY_TEAMPLATE = "SELECT count() FROM %(table_name)s FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'"
+INSERT_QUERY_TEMPLATE = """\
+INSERT INTO %(table_name)s (%(required_columns)s)
+SELECT %(select_columns)s
+FROM %(table_name)s FINAL
+WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'
+"""
+
+FINAL_QUERY_TEMPLATE = {
+    "required_columns": "project_id, timestamp, event_id",
+    "select_columns": "project_id, timestamp, event_id, group_id, primary_hash",
+}
+
+
+def _build_cluster(healthy: bool = True) -> FakeClickhouseCluster:
+    return FakeClickhouseCluster(
+        host="query_node",
+        port=9000,
+        user="default",
+        password="",
+        database="default",
+        http_port=8123,
+        storage_sets={"events"},
+        single_node=False,
+        cluster_name="my_cluster",
+        distributed_cluster_name="my_distributed_cluster",
+        nodes=[
+            ClickhouseNode("storage-0-0", 9000, 1, 1),
+            ClickhouseNode("storage-0-1", 9000, 1, 2),
+            ClickhouseNode("storage-1-0", 9000, 2, 1),
+            ClickhouseNode("storage-1-1", 9000, 2, 2),
+            ClickhouseNode("storage-2-0", 9000, 3, 1),
+            ClickhouseNode("storage-2-1", 9000, 3, 2),
+        ],
+        healthy=healthy,
+    )
+
+
+@pytest.mark.parametrize(
+    "write_node_replacements_projects, expected_queries", TEST_CASES
+)
+def test_write_single_node(
+    write_node_replacements_projects: str,
+    expected_queries: Mapping[str, Sequence[str]],
+) -> None:
+    """
+    Test the execution of replacement queries on both storage nodes and
+    query nodes.
+    """
+    set_config("write_node_replacements_projects", write_node_replacements_projects)
+    test_cluster = _build_cluster()
+    # Cannot use patch properly for this as I cannot patch the use case
+    # with a new instance for each parameter in the parametrized test.
+    current_clusters = cluster.CLUSTERS
```

Comment:
You don't have to change this, but this is something that could be done nicely with a pytest fixture.

Reply:
So I have two issues with that.
This test is also parametrized (@pytest.mark.parametrize). I always had issues when trying to use a fixture in a parametrized test. Did you manage to make it work somewhere?

Second, what do you suggest to replace with a fixture? I think I would still have to set a module scoped variable in `clsuters`. Line 132 and reset it at the end of the test.mOr are you suggesting to just replace the `_build_cluster` call with a fixture ?

Reply:
My thought was to take lines 130-133, and 154-155 and put them into a fixture of the form

```
@pytest.fixture
def override_cluster():
    current_clusters = cluster.CLUSTERS
    current_mapping = cluster._STORAGE_SET_CLUSTER_MAP
    
    def override(cluster, storage_map):
        clusters.CLUSTERS = [cluster]
        cluster._STORAGE_SET_CLUSTER_MAP = storage_map
        
    yield override
    
    cluster.CLUSTERS = current_clusters
    cluster._STORAGE_SET_CLUSTER_MAP = current_mapping
```

I think you could do this using an indirect parameterization (https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments) but I haven't actually tested that. I mentioned it just because you have this pattern repeated a few times.

Reply:
ok, will try.

Reply:
Found a way. It actually works

Reply:
Nice.